File: interfaces/julia/src/reactornet.jl
# ReactorNet: time integration of a network of reactors.

"""
    ReactorNet(reactors::Vector{Reactor})
    ReactorNet(reactor::Reactor)

Create a reactor network for time integration.  The network keeps references to
its reactors so they are not finalized while integration is in progress.
"""
mutable struct ReactorNet <: CanteraObject
    handle::Int32
    reactors::Vector{Reactor}
    closed::Bool

    function ReactorNet(reactors::Vector{Reactor})
        isempty(reactors) && throw(ArgumentError("ReactorNet needs at least one reactor"))
        handles = Int32[r.handle for r in reactors]
        h = check(LibCantera.reactornet_new(Int32(length(handles)), pointer(handles)))
        net = new(h, reactors, false)
        finalizer(close!, net)
        return net
    end
end

ReactorNet(r::Reactor) = ReactorNet([r])

function close!(net::ReactorNet)
    net.closed && return net
    net.closed = true
    try
        LibCantera.reactornet_del(net.handle)
    catch
    end
    return net
end

"""
    advance!(net, t)

Advance the network state to absolute time `t` [s].
"""
function advance!(net::ReactorNet, t)
    check(LibCantera.reactornet_advance(net.handle, Float64(t)))
    return net
end

"""
    step!(net) -> Float64

Take one internal timestep and return the new time [s].
"""
step!(net::ReactorNet) = checkd(LibCantera.reactornet_step(net.handle))

"Current network time [s].  Extends `Base.time` for `ReactorNet`."
Base.time(net::ReactorNet) = checkd(LibCantera.reactornet_time(net.handle))

"Set the initial integration time [s]."
function set_initial_time!(net::ReactorNet, t)
    check(LibCantera.reactornet_setInitialTime(net.handle, Float64(t)))
    return net
end

"Set the maximum internal timestep [s]."
function set_max_time_step!(net::ReactorNet, dt)
    check(LibCantera.reactornet_setMaxTimeStep(net.handle, Float64(dt)))
    return net
end

"""
    set_tolerances!(net; rtol=1e-9, atol=1e-15)

Set the relative and absolute integrator tolerances.
"""
function set_tolerances!(net::ReactorNet; rtol=1e-9, atol=1e-15)
    check(LibCantera.reactornet_setTolerances(net.handle, Float64(rtol), Float64(atol)))
    return net
end

"Relative error tolerance of the network integrator."
rtol(net::ReactorNet) = checkd(LibCantera.reactornet_rtol(net.handle))

"Absolute error tolerance of the network integrator."
atol(net::ReactorNet) = checkd(LibCantera.reactornet_atol(net.handle))

"""
    set_sensitivity_tolerances!(net; rtol=1e-6, atol=1e-6)

Set the relative and absolute tolerances used for sensitivity analysis.
"""
function set_sensitivity_tolerances!(net::ReactorNet; rtol=1e-6, atol=1e-6)
    check(LibCantera.reactornet_setSensitivityTolerances(net.handle,
                                                         Float64(rtol), Float64(atol)))
    return net
end

"""
    sensitivity(net, component, p, reactor) -> Float64

Normalized sensitivity of `component` (e.g. `"temperature"` or a species name)
in `reactor` with respect to sensitivity parameter `p` (1-based).  `reactor` may
be a [`Reactor`](@ref) belonging to the network or its 1-based position.
"""
function sensitivity(net::ReactorNet, component::AbstractString, p::Integer,
                     reactor::Reactor)
    idx = findfirst(r -> r === reactor, net.reactors)
    idx === nothing && throw(ArgumentError("reactor is not part of this network"))
    return sensitivity(net, component, p, idx)
end

function sensitivity(net::ReactorNet, component::AbstractString, p::Integer,
                     reactor::Integer)
    return checkd(LibCantera.reactornet_sensitivity(net.handle, component,
                                                    Int32(p - 1), Int32(reactor - 1)))
end

"""
    n_components(net) -> Int

Number of state variables (equations) integrated by the network.
"""
n_components(net::ReactorNet) = Int(check(LibCantera.reactornet_neq(net.handle)))

"""
    state(net) -> Vector{Float64}

Current network state vector, of length [`n_components`](@ref).
"""
state(net::ReactorNet) =
    get_array(n_components(net), (n, b) -> LibCantera.reactornet_getState(net.handle, n, b))

"Name of state-vector component `i` (1-based)."
function component_name(net::ReactorNet, i::Integer)
    return get_string((n, b) -> LibCantera.reactornet_componentName(net.handle, Int32(i - 1), n, b))
end

"""
    component_names(net) -> Vector{String}

Names of all state-vector components, aligned with [`state`](@ref).
"""
component_names(net::ReactorNet) = [component_name(net, i) for i in 1:n_components(net)]

Base.show(io::IO, net::ReactorNet) =
    print(io, net.closed ? "ReactorNet(<closed>)" :
          "ReactorNet($(length(net.reactors)) reactor(s), t=$(time(net)) s)")

File: interfaces/matlab/+ct/+zeroD/ReactorNet.m
classdef ReactorNet < handle
    % ReactorNet class ::
    %
    %     >> r = ct.zeroD.ReactorNet(reactors)
    %
    % A :mat:class:`ct.zeroD.ReactorNet` object is a container that holds one or more
    % :mat:class:`ct.zeroD.ReactorBase` objects in a network. :mat:class:`ct.zeroD.ReactorNet`
    % objects are used to simultaneously advance the state of one or more coupled
    % reactors.
    %
    % :param reactors:
    %    An instance of or a cell array of instances of class
    %    :mat:class:`ct.zeroD.ReactorBase`.

    properties (SetAccess = immutable)
        id = -1

        % Cell array of the :mat:class:`ct.zeroD.ReactorBase` objects that make up
        % the network, in the order in which they were installed.
        reactors
    end

    properties (SetAccess = public)

        % Current time [s].
        % The setter method sets the time at which integration should be
        % restarted, using the current state as the initial condition.
        time

        % Max time step [s].
        %
        % The integrator chooses a step size based on the desired error
        % tolerance and the rate at which the solution is changing.
        % In some cases, the solution changes very slowly at first,
        % then very rapidly (ignition problems). In such cases, the
        % integrator may choose a timestep that is too large, which
        % leads to numerical problems later. Use this method to set an
        % upper bound on the timestep.
        maxTimeStep

        atol % Absolute error tolerance.

        rtol % Relative error tolerance.

    end

    methods
        %% ReactorNet Class Constructor

        function obj = ReactorNet(reactors)
            arguments
                reactors
            end
            if isa(reactors, 'ct.zeroD.ReactorBase')
                % Allow simpler syntax for creating a network with one reactor.
                reactors = {reactors};
            end
            reactorIDs = cellfun(@(r) r.id, reactors);

            obj.id = ct.impl.call('mReactornet_new', reactorIDs);
            obj.reactors = reactors;
            obj.time = 0;

        end

        %% ReactorNet Class Destructor

        function delete(obj)
            % Delete the :mat:class:`ct.zeroD.ReactorNet` object object.
            if obj.id >= 0
                ct.impl.call('mReactornet_del', obj.id);
            end
        end

        %% ReactorNet Utility Methods

        function advance(obj, tout)
            % Advance the state of the reactor network in time. ::
            %
            %     >> n.advance(tout)
            %
            % Method `advance` integrates the system of ODEs that determine
            % the rate of change of the volume, the mass of each species,
            % and the total energy for each reactor. The integration is
            % carried out from the current time to "tout". (Note: "tout" is
            % an absolute time, not a time interval.) The integrator may
            % take many internal timesteps before reaching tout.
            %
            % :param tout:
            %    End time [s] of the integration.

            ct.impl.call('mReactornet_advance', obj.id, tout);
        end

        function t = step(obj)
            % Take a single internal step. ::
            %
            %     >> n.step()
            %
            % The time/distance after taking the step is returned.

            t = ct.impl.call('mReactornet_step', obj.id);
        end

        %% ReactorNet set methods

        function set.time(obj, t)
            ct.impl.call('mReactornet_setInitialTime', obj.id, t);
        end

        function set.maxTimeStep(obj, maxstep)
            % Set the maximum time step. ::
            %
            %     >> n.setMaxTimeStep(maxstep)
            %
            % The integrator chooses a step size based on the desired error
            % tolerance and the rate at which the solution is changing.
            % In some cases, the solution changes very slowly at first,
            % then very rapidly (ignition problems). In such cases, the
            % integrator may choose a timestep that is too large, which
            % leads to numerical problems later. Use this method to set an
            % upper bound on the timestep.
            %
            % :param maxstep:
            %    max time step [s].

            ct.impl.call('mReactornet_setMaxTimeStep', obj.id, maxstep);
        end

        function setSensitivityTolerances(obj, rerr, aerr)
            % Set the error tolerance for sensitivity analysis. ::
            %
            %     >> n.setSensitivityTOlerances(nerr, aerr)
            %
            % :param rerr:
            %    Scalar relative error tolerance.
            % :param aerr:
            %    Scalar absolute error tolerance.

            ct.impl.call('mReactornet_setSensitivityTolerances', obj.id, rerr, aerr);
        end

        function set.atol(obj, aerr)
            ct.impl.call('mReactornet_setAbsoluteTolerance', obj.id, aerr);
        end

        function set.rtol(obj, rerr)
            ct.impl.call('mReactornet_setRelativeTolerance', obj.id, rerr);
        end

        %% ReactorNet get methods

        function t = get.time(obj)
            t = ct.impl.call('mReactornet_time', obj.id);
        end

        function t = get.atol(obj)
            t = ct.impl.call('mReactornet_atol', obj.id);
        end

        function t = get.rtol(obj)
            t = ct.impl.call('mReactornet_rtol', obj.id);
        end

        function s = sensitivity(obj, component, p, r)
            % Sensitivity of a solution variable with respect to a sensitivity
            % parameter. ::
            %
            %     >> s = n.sensitivity(component, p)
            %     >> s = n.sensitivity(component, p, r)
            %
            % :param component:
            %    Name of the solution variable, given as a character array or a
            %    string, for example ``'temperature'`` or a species name. It is
            %    resolved within the reactor `r`. Alternatively, an integer giving
            %    the 1-based index of the variable in the global state vector of
            %    the network, in which case `r` is not used.
            % :param p:
            %    Integer, 1-based index of the sensitivity parameter.
            % :param r:
            %    The reactor in which `component` is to be found, given either as an
            %    instance of :mat:class:`ct.zeroD.ReactorBase` belonging to this
            %    network or as its 1-based position within the network. Defaults to
            %    the first reactor in the network.
            % :return:
            %    Scalar normalized sensitivity coefficient.

            arguments
                obj
                component
                p (1,1) double
                r = 1
            end

            if isnumeric(component)
                % Global index into the network state vector
                s = ct.impl.call('mReactornet_sensitivity', obj.id, ...
                                 component - 1, p - 1);
                return
            end

            if ~(ischar(component) || isstring(component))
                error(['Sensitivity component must be a character array, a ' ...
                       'string, or an integer index, not a %s.'], class(component));
            end

            if isa(r, 'ct.zeroD.ReactorBase')
                index = find(cellfun(@(x) x == r, obj.reactors), 1);
                if isempty(index)
                    error('The given reactor is not part of this network.');
                end
            elseif isnumeric(r) && isscalar(r)
                index = r;
            else
                error(['Reactor must be a ct.zeroD.ReactorBase object or a ' ...
                       '1-based index, not a %s.'], class(r));
            end

            s = ct.impl.call('mReactornet_sensitivityByName', obj.id, ...
                             char(component), p - 1, index - 1);

        end

    end

end
